In [2]:
import pandas as pd

In [4]:
df_train = pd.read_csv("train_KNN.csv")
df_test = pd.read_csv("test_KNN.csv")

In [5]:
df_train.head()
df_test.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,0,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,1,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,0,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,0,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,1,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


In [6]:
df_train.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


In [7]:
mp = 100*(df_train.isnull().sum()/len(df_train))
mp

,0
PassengerId,0.000000
Survived,0.000000
Pclass,0.000000
Name,0.000000
Sex,0.000000
Age,19.865320
SibSp,0.000000
Parch,0.000000
Ticket,0.000000
Fare,0.000000


In [8]:
mp = mp[mp>0].sort_values()
mp

,0
Embarked,0.224467
Age,19.865320
Cabin,77.104377


In [9]:
df_train["Age"] = df_train["Age"].fillna(df_train["Age"].median())
df_train = df_train.drop(["Cabin"], axis=1)
df_train = df_train.dropna(axis=0, subset=["Embarked"])

In [10]:
df_test["Age"] = df_test["Age"].fillna(df_test["Age"].median())
df_test = df_test.drop(["Cabin"], axis=1)
df_test = df_test.dropna(axis=0, subset=["Fare"])

In [11]:
df_train_num = df_train.select_dtypes(exclude="object")
df_train_obj = df_train.select_dtypes(include="object").drop(["Name", "Ticket"], axis=1)

df_test_num = df_test.select_dtypes(exclude="object")
df_test_obj = df_test.select_dtypes(include="object").drop(["Name", "Ticket"], axis=1)

In [12]:
df_train_obj = pd.get_dummies(df_train_obj, drop_first=True)
df_test_obj = pd.get_dummies(df_test_obj, drop_first=True)

In [17]:
df_test_obj = df_test_obj.reindex(columns=df_train_obj.columns, fill_value=0)

In [13]:
Final_train_df = pd.concat([df_train_num, df_train_obj], axis=1)
Final_test_df = pd.concat([df_test_num, df_test_obj], axis=1)

In [14]:
X_train = Final_train_df.drop("Survived", axis=1)
y_train = Final_train_df["Survived"]

X_test = Final_test_df.drop("Survived", axis=1)
y_test = Final_test_df["Survived"]

In [20]:
X_test = X_test.reindex(columns=X_train.columns, fill_value=0)

In [21]:
from sklearn.preprocessing import StandardScaler

# FIX
X_test = X_test.reindex(columns=X_train.columns, fill_value=0)

scaler = StandardScaler()
scaler.fit(X_train)

scaled_X_train = scaler.transform(X_train)
scaled_X_test = scaler.transform(X_test)

In [22]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(scaled_X_train, y_train)

y_pred = knn.predict(scaled_X_test)

print("KNN Accuracy:", accuracy_score(y_test, y_pred))

KNN Accuracy: 0.3764988009592326


In [23]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()
lr.fit(scaled_X_train, y_train)

y_pred_lr = lr.predict(scaled_X_test)

print("Logistic Accuracy:", accuracy_score(y_test, y_pred_lr))

Logistic Accuracy: 0.7338129496402878


In [24]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

print("Decision Tree Accuracy:", accuracy_score(y_test, y_pred_dt))

Decision Tree Accuracy: 0.7793764988009593


## Final Interpretation

In this experiment, the Titanic dataset was analysed using three different machine learning models: K-Nearest Neighbors (KNN), Logistic Regression, and Decision Tree Classifier. Before applying the models, data preprocessing was performed, which included handling missing values, removing irrelevant features, encoding categorical variables using dummification, and standardizing numerical features.

The KNN model produced an accuracy of 37.65%, which is significantly low compared to the other models. This indicates that KNN was not suitable for this dataset, possibly due to its sensitivity to feature scaling and high dimensionality after encoding.

The Logistic Regression model achieved an accuracy of 73.38%, showing a considerable improvement. This suggests that the dataset has a relatively linear relationship between the input features and the target variable.

The Decision Tree Classifier performed the best with an accuracy of 77.94%. This model was able to capture complex patterns and interactions within the dataset more effectively than the other models.

Overall, it can be concluded that the Decision Tree model is the most suitable for this dataset.